### PR TITLE
Use iterator instead of generator in ListWalker for better performance.

### PR DESCRIPTION
**Pull request by filipsobol**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Other (list): Use iterator instead of generator in ListWalker for better performance.

---

The generator introduces some unnecessary overhead. This PR makes the `lists` performance demo 9% faster.

### CODE REVIEW

File: packages/ckeditor5-list/src/list/utils/listwalker.ts [hunk 1]
```diff
@@ -191,47 +191,69 @@ export default class ListWalker {
...
-): IterableIterator<ListIteratorValue> {
-	const isForward = direction == 'forward';
-	const previousNodesByIndent: Array<ListElement> = []; // Last seen nodes of lower indented lists.
-	let previous = null;
+export class SiblingListBlocksIterator implements IterableIterator<ListIteratorValue> {
+	private node: Node | null;
+	private isForward: boolean;
+	private previousNodesByIndent: Array<ListElement> = [];
+	private previous: ListElement | null = null;
+	private previousNodeIndent: number | null = null;
```
Comment (niegowski) [inline packages/ckeditor5-list/src/list/utils/listwalker.ts:200]:
```suggestion
	private _node: Node | null;
	private _isForward: boolean;
	private _previousNodesByIndent: Array<ListElement> = [];
	private _previous: ListElement | null = null;
	private _previousNodeIndent: number | null = null;
```

Comment (filipsobol) [reply packages/ckeditor5-list/src/list/utils/listwalker.ts]: I forget about this every time 😮‍💨

Fixed

File: packages/ckeditor5-list/src/list/utils/listwalker.ts [hunk 2]
```diff
@@ -191,47 +191,69 @@ export default class ListWalker {
...
 
-		yield { node, previous, previousNodeInList };
+		const value = {
+			node: this._node,
+			previous: this._previous,
+			previousNodeInList
+		};
+
+		this._previous = this._node as ListElement;
+		this._previousNodeIndent = nodeIndent;
```
Comment (scofalik) [inline packages/ckeditor5-list/src/list/utils/listwalker.ts:253]: It's not easy to compare these changes so forgive me if I am making a mistake somewhere here, but.

After first iteration `nodeIndent` is `null`, so `_previousNodeIndent` will be also `null`. In second iteration, in line 228, `previousNodeIndent` will be set to `null`. I think this is different than before.

Comment (scofalik) [reply packages/ckeditor5-list/src/list/utils/listwalker.ts:253]: Other than it looks to me that the iteration logic is the same.

Comment (filipsobol) [reply packages/ckeditor5-list/src/list/utils/listwalker.ts:252]: You're right, fixed.

**Review by scofalik — APPROVED** [review]

**Pull request merged**